### ISSUE TITLE
Brew auto-changes branch when developing formula

### ISSUE THREAD

**Issue by sigmavirus24** [Homebrew/brew#737]
# Please follow the general troubleshooting steps first:

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [x] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?

### Bug reports:

I was developing an update to formula in homebrew-core and when I tried to develop it, brew auto-switched back to master. I talked to @tdsmith and he indicated I should have set `HOMEBREW_DEVELOPER=1` to bypass this new behaviour.

In short, this was surprising. I understand there was no way to warn about this, but it would have useful to know about. Perhaps we can put it homebrew-core's README/CONTRIBUTING/TROUBLESHOOTING documentation.

**Comment by maxnordlund** [comment]
One other thing, it would be nicer if it poped the changes/ignored dirty taps, since then you are most likely working on something. Or just warn that it is dirty/not on master.

**Comment by MikeMcQuaid** [comment]
It does this if you have `HOMEBREW_DEVELOPER` set. I'll provide a decent response to this beyond that workaround in future. The short version: the hard thing here is getting the balance for Homebrew developers vs. those who don't know what Git is or really understand how to use it.

**Comment by MikeMcQuaid** [comment]
Have opened a PR for this in https://github.com/Homebrew/brew/pull/881.
